### DISCUSSION TITLE
4. TCP와 UDP의 차이점에 대해서 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#2] [General]
TCP와 UDP의 차이점에 대해서 설명해 주세요.

**Comment by eun-byeol** [comment]
### 🔑 키워드
TCP - `연결 지향형 프로토콜`, `신뢰성 보장`
UDP - `비연결 지향형 프로토콜`, `빠른 속도`

---

TCP는 `연결 지향형 프로토콜`로, 데이터 전송 전 송수신자 간 연결을 설정하여 `신뢰성을 보장`합니다. UDP는 `비연결 지향형 프로토콜`로 연결 설정 없이 데이터를 전송하여 `속도가 빠르지`만, 신뢰성은 보장되지 않습니다.

TCP는 `흐름제어, 혼잡제어, 오류제어`를 통해 신뢰성을 보장합니다. UDP는 이런 절차가 없고, `더 간단한 헤더 구조`로 오버헤드 줄이고 속도가 빠릅니다. UDP는 신뢰성을 보장하진 않지만, 헤더의 checksum 필드로 최소한의 오류 검출을 할 수 있습니다.

TCP는 1:1 통신(송수신자 간 통신)만 수행할 수 있고, UDP는 1:1, 1:N, N:M 통신이 가능합니다.

  | TCP(Transmission Control Protocol) | UDP(User Datagram Protocol)
-- | -- | --
연결 방식 | 연결형(3-way handshake를 통한 연결 설정) | 비연결형(데이터그램* 방식)
신뢰성 | 높음 | 낮음
속도 | 느림 | 빠름
전송 순서 보장 여부 | O | X
수신 확인 여부 | O | X
통신 방식 | 1:1 | 1:1, 1:N, N:M

---
### 더 알아보기
> 데이터그램

비연결형 네트워크에서 데이터를 전달하기 위해 사용되는 `독립적인 데이터 단위`

> 네트워크 통신 방식

- 유니캐스트(1:1)
    ![image](https://github.com/user-attachments/assets/44c3b8f7-cf47-4e03-b8a9-a9859f8791a7)
    
- 멀티캐스트(1:N) : 네트워크에 연결된 시스템 중, `특정 그룹`에 속해 있는 시스템에게만 데이터를 전송하는 방식
    ![image](https://github.com/user-attachments/assets/b114ca5d-32e3-4e06-98d9-913aa3d5a769)
    
- 브로드캐스트 : 네트워크에 연결된 `모든 시스템`에게 데이터 전송하는 방식
    ![image](https://github.com/user-attachments/assets/2455d295-aac9-49c1-931f-fe7561e6427e)
